dcrec/secp256k1/field.go

```diff
@@ -894,6 +894,165 @@ func (f *fieldVal) Mul2(val *fieldVal, val2 *fieldVal) *fieldVal {
+	a176.Square().Square().Square().Square().Square()         // a176 = a^(2^138 - 2^50)
+	a176.Square().Square().Square().Square().Square()         // a176 = a^(2^143 - 2^55)
+	a176.Square().Square().Square().Square().Square()         // a176 = a^(2^148 - 2^60)
+	a176.Square().Square().Square().Square().Square()         // a176 = a^(2^153 - 2^65)
+	a176.Square().Square().Square().Square().Square()         // a176 = a^(2^158 - 2^70)
+	a176.Square().Square().Square().Square().Square()         // a176 = a^(2^163 - 2^75)
+	a176.Square().Square().Square().Square().Square()         // a176 = a^(2^168 - 2^80)
+	a176.Square().Square().Square().Square().Square()         // a176 = a^(2^173 - 2^85)
+	a176.Square().Square().Square()                           // a176 = a^(2^176 - 2^88)
+	a176.Mul(&a88)                                            // a176 = a^(2^176 - 1)
+	a220.SquareVal(&a176).Square().Square().Square().Square() // a220 = a^(2^181 - 2^5)
+	a220.Square().Square().Square().Square().Square()         // a220 = a^(2^186 - 2^10)
+	a220.Square().Square().Square().Square().Square()         // a220 = a^(2^191 - 2^15)
+	a220.Square().Square().Square().Square().Square()         // a220 = a^(2^196 - 2^20)
+	a220.Square().Square().Square().Square().Square()         // a220 = a^(2^201 - 2^25)
+	a220.Square().Square().Square().Square().Square()         // a220 = a^(2^206 - 2^30)
+	a220.Square().Square().Square().Square().Square()         // a220 = a^(2^211 - 2^35)
+	a220.Square().Square().Square().Square().Square()         // a220 = a^(2^216 - 2^40)
+	a220.Square().Square().Square().Square()                  // a220 = a^(2^220 - 2^44)
+	a220.Mul(&a44)                                            // a220 = a^(2^220 - 1)
+	a223.SquareVal(&a220).Square().Square()                   // a223 = a^(2^223 - 2^3)
+	a223.Mul(&a3)                                             // a223 = a^(2^223 - 1)
+
+	f.SquareVal(&a223).Square().Square().Square().Square() // f = a^(2^228 - 2^5)
+	f.Square().Square().Square().Square().Square()         // f = a^(2^233 - 2^10)
+	f.Square().Square().Square().Square().Square()         // f = a^(2^238 - 2^15)
+	f.Square().Square().Square().Square().Square()         // f = a^(2^243 - 2^20)
+	f.Square().Square().Square()                           // f = a^(2^246 - 2^23)
+	f.Mul(&a22)                                            // f = a^(2^246 - 2^22 - 1)
+	f.Square().Square().Square().Square().Square()         // f = a^(2^251 - 2^27 - 2^5)
+	f.Square()                                             // f = a^(2^252 - 2^28 - 2^6)
+	f.Mul(&a2)                                             // f = a^(2^252 - 2^28 - 2^6 - 2^1 - 1)
+	f.Square().Square()                                    // f = a^(2^254 - 2^30 - 2^8 - 2^3 - 2^2)
+	//                                                     //   = a^(2^254 - 2^30 - 244)
+	//                                                     //   = a^((p+1)/4)
```

Comment:
Ok, I spent waaaaay longer on this than I should have and I haven't internalized where this specific set operations came from. But I've verified that all the provided equations hold via wolfram alpha and it computes what it claims to.

`[a3] a^(2^3-1) == (a^(2^2-1))^2 * a`
https://www.wolframalpha.com/input/?i=a%5E%282%5E3-1%29+%3D%3D+%28a%5E%282%5E2-1%29%29%5E2+*+a

`[a6] a^(2^6-1) = (a^(2^3-1))^(2^3) * a^(2^3-1)`
https://www.wolframalpha.com/input/?i=a%5E%282%5E6-1%29+%3D+%28a%5E%282%5E3-1%29%29%5E%282%5E3%29+*+a%5E%282%5E3-1%29

`[a9] a^(2^9-1) = (a^(2^6-1))^(2^3) * a^(2^3-1)`
https://www.wolframalpha.com/input/?i=a%5E%282%5E9-1%29+%3D+%28a%5E%282%5E6-1%29%29%5E%282%5E3%29+*+a%5E%282%5E3-1%29

`[a11] a^(2^11-1) = (a^(2^9-1))^(2^2) * a^(2^2-1)`
https://www.wolframalpha.com/input/?i=a%5E%282%5E11-1%29+%3D+%28a%5E%282%5E9-1%29%29%5E%282%5E2%29+*+a%5E%282%5E2-1%29

`[a22] a^(2^22-1) = (a^(2^11-1))^(2^11) * a^(2^11-1)`
https://www.wolframalpha.com/input/?i=a%5E%282%5E22-1%29+%3D+%28a%5E%282%5E11-1%29%29%5E%282%5E11%29+*+a%5E%282%5E11-1%29

`[a44] a^(2^44-1) = (a^(2^22-1))^(2^22) * a^(2^22-1)`
https://www.wolframalpha.com/input/?i=a%5E%282%5E44-1%29+%3D+%28a%5E%282%5E22-1%29%29%5E%282%5E22%29+*+a%5E%282%5E22-1%29

`[a88] a^(2^88-1) = (a^(2^44-1))^(2^44) * a^(2^44-1)`
https://www.wolframalpha.com/input/?i=a%5E%282%5E88-1%29+%3D+%28a%5E%282%5E44-1%29%29%5E%282%5E44%29+*+a%5E%282%5E44-1%29

`[a176] a^(2^176-1) = (a^(2^88-1))^(2^88) * a^(2^88-1)`
https://www.wolframalpha.com/input/?i=a%5E%282%5E176-1%29+%3D+%28a%5E%282%5E88-1%29%29%5E%282%5E88%29+*+a%5E%282%5E88-1%29

`[a220] a^(2^220-1) = (a^(2^176-1))^(2^44) * a^(2^44-1)`
https://www.wolframalpha.com/input/?i=a%5E%282%5E220-1%29+%3D+%28a%5E%282%5E176-1%29%29%5E%282%5E44%29+*+a%5E%282%5E44-1%29

`[a223] a^(2^223-1) = (a^(2^220-1))^(2^3) * a^(2^3-1)`
https://www.wolframalpha.com/input/?i=a%5E%282%5E223-1%29+%3D+%28a%5E%282%5E220-1%29%29%5E%282%5E3%29+*+a%5E%282%5E3-1%29

`[a(p+1/4)] a^(2^254-2^30-244) = ((((a^(2^223-1))^(2^23) * a^(2^22-1))^(2^6)) * a^(2^2-1))^2^2`
https://www.wolframalpha.com/input/?i=a%5E%282%5E254-2%5E30-244%29+%3D+%28%28%28%28a%5E%282%5E223-1%29%29%5E%282%5E23%29+*+a%5E%282%5E22-1%29%29%5E%282%5E6%29%29+*+a%5E%282%5E2-1%29%29%5E2%5E2

Reply:
The intuition is that it's just summing the partial products.  Or stated differently, it's breaking up the multiplication into sum of easier multiplications.

For example, imagine you have the number 123045067 in decimal.  It's essentially just doing 123*1000000 + 45*1000 + 67*1 = 123045067, except in base 2 and with much much larger numbers.